Implement AES Encryption for Voice

viztea opened the pull request:
AES for voice receive is currently unsupported. 

I tested voice send/receive for all 3 xsalsa20 poly1305 nonce strategies, all work as expected.

DRSchlaubi reviewed (changes requested):
please run `./gradlew apiDump`

lukellmann commented:
i'm not very familiar with voice connections, can you explain a bit why this is needed?

viztea commented:
> i'm not very familiar with voice connections, can you explain a bit why this is needed?

It kinda is and kinda isn't. I initially wanted to implement AES because it was faster than xsalsa but after doing a couple benchmarks it looks like xsalsa is a little bit more consistent and a couple microseconds faster.

This isn't really that surprising considering Kord Voice isn't the most efficient voice library, it doesn't help that Java relies on `ByteBuffer`s to be performant.

lost-illusi0n commented on voice/src/main/kotlin/VoiceConnectionBuilder.kt:
```diff
@@ -166,19 +167,23 @@ public class VoiceConnectionBuilder(
         val udpSocket = udpSocket ?: GlobalVoiceUdpSocket
         val audioProvider = audioProvider ?: EmptyAudioPlayerProvider
-        val nonceStrategy = nonceStrategy ?: LiteNonceStrategy()
+        val encryption = if ((receiveVoice || streams != null) && encryption?.supportsDecryption != true) {
+            VoiceEncryption.XSalsaPoly1305()
+        } else {
+            encryption ?: VoiceEncryption.AeadAes256Gcm
+        }
```
It might be useful to log if the user's chosen encryption method ends up not being used.

lukellmann replied:
or even throw an exception? but logging a warning is probably best.

lost-illusi0n commented:
Functionality-wise LGTM. It likely will make adding/working with encryption a bit easier in the future too, if (or rather when) Discord decides to change things up.

lukellmann commented on voice/src/main/kotlin/encryption/VoiceEncryption.kt:
```diff
@@ -0,0 +1,163 @@
+
+    public val nonceLength: Int
+
+    public val supportsDecryption: Boolean get() = true
```
not having a default implementation here will avoid mistakes if another `VoiceEncryption` is added at some point

lukellmann commented on voice/src/main/kotlin/encryption/strategies/NonceStrategy.kt:
```diff
@@ -27,4 +23,17 @@ public sealed interface NonceStrategy {
      */
     public fun append(nonce: ByteArrayView, cursor: MutableByteArrayCursor)
+
+    public interface Factory {
```
why do we need this factory? i'd rather not add it if not needed.

lukellmann commented on voice/src/main/kotlin/encryption/VoiceEncryption.kt:
```diff
@@ -0,0 +1,163 @@
+            get() = false
+
+        override fun createBox(key: ByteArray): Box = object : Box {
+            //
```
what are these empty comments for?

viztea commented:
- Implemented voice receive for AES, had to refactor a bit of the code to support passing of AEAD data.
- Removed `VoiceEncryption#supportsDecryption`
- Removed the sealed modifier for `VoiceEncryption` to allow external implementations (??)
- Added kdoc to the `VoiceEncryption` interface.

Not sure if removing supportsDecryption is a good idea with it being an open interface 🤷🏼
